build: Add better SQL Test Infrastructure

manticore-projects opened the pull request:
- Switch to JUnit5 for parametrized tests and parallel execution
- Maven: Add Airlist Check SKIP switches for fast prototyping
- ~~Maven: remove Checklist Modules, since those are provided by Airbase~~ turns out, those are needed r/ Java 8
- Eclipse formatter configuration for Airlift style
- SQL Parser and Unparser Tests, read from SQL Files
- Updated some Tests to benefit from JUnit5
- Sphinx Website for deployment Github Pages
- Readme updated
- Added Gradle Build support
- Updated Airlift/Base to io.airlift ~~219~~ 100 (Last Java 8 supporting version)
- Provide explicit Airlift Checkstyle config
- Move the JavaCC Grammar to a more appropriate place
- Generate RailRoad Diagrams for the Grammar
- Downgrade Maven JUnit5 as requested by Airlift
- Add an ASCII Tree for visualizing the AST

manticore-projects committed:
build: Add better SQL Test Infrastructure

- Switch to JUnit5 for parametrized tests and parallel execution
- Maven: Add Airlist Check SKIP switches for fast prototyping
- Maven: remove Checklist Modules, since those are provided by Airbase
- Eclipse formatter configuration for Airlift style
- SQL Parser and Unparser Tests, read from SQL Files
- Updated some Tests to benefit from JUnit5

manticore-projects committed:
build: Add Spotless Plugin

Maven Spotless Plugin to enforce Airlift compliant formatting based on Eclipse formatter

manticore-projects committed:
doc: Add basic HTML documentation based on Sphinx

Add basic Sphinx Template with minimal content
Add Gradle task for Changelog
Add Gradle task for Rail Road Diagrams (although it does not work yet because this Grammar is strong)
Add Gradle task for building the Sphinx documentation, including the Prologue
Add Github Action for deploying the HTML to Github Pages
Add Github Action for Maven build
Add Github Issue template
Tweak the .gitignore file

manticore-projects committed:
fix: Git Branch is MAIN

manticore-projects committed:
fix: Try using native GIT

manticore-projects committed:
doc: Update the readme

manticore-projects commented:
Greetings.

I have worked a bit around the infrastructure. Added some Readme, Website/Github Pages and Test SQLs (so to read the statements from a formatted SQL file instead from Java Code).

I also wanted to add Rail-Road Diagrams, but this did not work yet since the Grammar uses some features I have never seen before. Bug reported https://github.com/GuntherRademacher/rr/issues/20

Added some support for formatting.

Please do let me know what you like or don't like.
At the moment some of the links may be wrong. I can fix that only when I know what you will accept and what the final location is.

Previews:
https://manticore-projects.github.io/sql/
https://github.com/manticore-projects/sql/tree/test_infrastructure

Cheers

kaikalur commented on parser/src/test/java/com/facebook/coresql/parser/TestSqlParser.java:
```diff
@@ -26,56 +32,34 @@
+    public static Stream<String> sqlStrings()
+    {
+        List<String> sqlStrings = Arrays.asList("use a.b;", " SELECT 1;", "SELECT a FROM T;", "SELECT a FROM T WHERE p1 > p2;", "SELECT a, b, c FROM T WHERE c1 < c2 and c3 < c4;", "SELECT CASE a WHEN IN ( 1 ) THEN b ELSE c END AS x, b, c FROM T WHERE c1 < c2 and c3 < c4;", "SELECT T.* FROM T JOIN W ON T.x = W.x;", "SELECT NULL;", "SELECT ARRAY[x] FROM T;", "SELECT TRANSFORM(ARRAY[x], x -> x + 2)  ...
+        return sqlStrings.stream();
```
Why did you concatenate these? Leave them on separate lines for ease of testing,

manticore-projects replied:
You are right of course. The Formatter was not tuned well enough yet.
I have addressed that with Commit https://github.com/prestodb/sql/pull/31/commits/24ddd02e68c8658c7285302de9be6c65dc4d59a3

Now the Formatter enforces Airlift rules and proper Line Wrapping for Arrays and Function Parameters.

kaikalur commented on parser/src/test/java/com/facebook/coresql/parser/TestSqlParser.java:
```diff
@@ -26,56 +32,34 @@
-        for (String sql : TEST_SQL_TESTSTRINGS) {
-            assertNotNull(parse(sql));
-        }
+        Assertions.assertNotNull(parse(sqlStr));
```
When you are at it, please add the failed sqlStr to the assert message

manticore-projects replied:
Done.

kaikalur commented on rewriter/src/test/java/com/facebook/coresql/rewriter/TestOrderByRewriter.java:
```diff
@@ -16,93 +16,41 @@
-                    " CAST(MAP() AS map<bigint,array<boolean>>) AS \"bool_tensor_features\";",
-            "SELECT f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f(f())))))))))))))))))))))))))))));",
-            "SELECT abs, 2 as abs;",
+            "SELECT * FROM T ORDER BY y;", "SELECT * FROM T ORDER BY y LIMIT 10;", "use a.b;", " SELECT 1;", "SELECT a FROM T;", "SELECT a FROM T WHERE p1 > p2;", "SELECT a, b, c FROM T WHERE c1 < c2 and c3 < c4;", "SELECT CASE a WHEN IN ( 1 ) THEN b ELSE c END AS x, b, c FROM T WHERE c1 < c2 and c3 < c4;", "SELECT T.* FROM T JOIN W ON T.x = W.x;", "SELECT NULL;", "SELECT ARRAY[x] FROM T;", "SELE ...
```
Same comment here - don't concatenate the test sql strings

manticore-projects replied:
Solved via Commit https://github.com/prestodb/sql/pull/31/commits/24ddd02e68c8658c7285302de9be6c65dc4d59a3

kaikalur commented on sphinx/src/keywords.rst:
```diff
@@ -0,0 +1,231 @@
+Restricted Keywords
+***********************
+
+The following Keywords are **restricted** in Version |JSQLPARSER_VERSION| and must not be used for **Naming Objects**: 
```
Why are we having references to JSQL here? I thought JSQL would simply use this package

manticore-projects replied:
Thanks for feedback.

This particular reference is only a variable name used in a template. It will be replaced/filled with the actual version.

Before changing it, we will want to agree on a proper Brand Name for this software and the decision is all yours. Will it be `Presto SQL Parser`, since it is linked to Presto?

Or should it aim broader, like `JSQL:2016', 'JavaSQL16', 'ParseSQL16', 'JSQLSuite', 'JSQLTools'?

Since we are on it, the Git Repo checks out as `sql` (which immediately collided with my  curated collection of the worst SQL statements I have seen in my life.)

Please make a choice and I would adjust everything accordingly.

kaikalur replied:
Hmm this is a tricky one. Definitely let's not call it jsql especially since we have C++ version (which your changes will hopefully not impact). Presto SQL parser might be most appropriate

manticore-projects replied:
So be it! I would adjust this tomorrow, also remove any manticore references and submit.
Thanks for the feedback!

I did not touch any C++  Version.

Although: I would like to have a comprehensive Test/CI Suite, which can warn us. If it succeeds, we should be good. When not, then reject.

manticore-projects committed:
style: Tune the Eclipse Formatter Configuration

Update the Line Wrapping Options for Function and Array Parameters.
Tweak the Eclipse Formatter Configuration to match the Airlift Checkstyle rules precisely.

Now, calling `mvn spotless:apply` will guarantee compliance with Airlift style.

manticore-projects committed:
test: Add many more Real Life SQL statements

Add plenty of SQL samples
Separate succeeding samples from failing samples so that failures will not block the build
Let the tests show the Statement and the Parser Error
Further tune the Eclipse Formatter since Airlift is really strict
Simplify some Regular Expressions

manticore-projects commented:
I would like to apologize upfront for the massive change-set.
However, I feel like we achieve something:

1) We have now a kind of SQL-File Test Cases, which DO PASS and should always pass -- whatever we do. `mvn package` would fail when those break.
2) We have an additional set of FAIL SAFE SQL-Files, which illustrate missing features of the parser. Those will show only with `mvn integration-test`.

Now we can pick from here one by one, analyze what goes wrong and decide if we want to support it or not.

Any new success should be moved to the mandatory test suite.
I like the fact, that the Tests can be provided as properly formatted SQL and so are easy to read. 

All tests are parseable with JSQLParser and have been formatted with JSQLFormatter.
I would like to suggest that as a first realistic goal (minus a few things like old Oracle joins or multiple column drop statements).

manticore-projects commented:
Illustration of the Test Suite:

![image](https://user-images.githubusercontent.com/18080123/209671044-f7b15908-0009-43dc-a0f0-78b52faf7546.png)

manticore-projects committed:
build: Gradle Build and Rail Road Diagrams

Added Gradle Build support
Updated Airlift/Base to io.airlift 219
Provide explicit Airlift Checkstyle config
Move the JavaCC Grammar to a more appropriate place
Generate RailRoad Diagrams for the Grammar
Downgrade Maven JUnit5 as requested by Airlift
Add an ASCII Tree for visualizing the AST

manticore-projects committed:
doc: changelog.rst

Signed-off-by: Andreas Reichel <andreas@manticore-projects.com>

kaikalur commented:
It maybe ok for now to have this large PR as it's just adding a bunch of new things and doesn't seem to change the grammar. But in general we prefer separate small PRs .

Also we generally squash all commits into one before merging but looks like you have many different things so it *maybe* ok. Also looks like this repo is also under Linux Foundation.  We need to check if it's ok to have manticore as a package under sql. I suggest creating a sql-manticore toplevel package so the grammar module is clean with just the grammar. So we should eventually have parser, linter, rewriter and formatter as the toplevel packages.

manticore-projects commented:
Happy new year 2023.

I know, that this PR is way too large -- however, it is really hard to work on Build/Test infrastructure and Documentation and keep it small, when the PR is kept pending for a long time.

I can remove any "manticore" related reference in the packages of course, it was just copy'n paste.

Btw, Rail Road Diagrams:  https://manticore-projects.github.io/sql/syntax.html

kaikalur commented:
It's fine to have manticore references but keep that in a separate package. I just want to keep the parser package super clean with no specific reference to anything else. Just like presto has a lot of connectors that all sit in presto-xxx packages.

kaikalur commented:
> Btw, Rail Road Diagrams: https://manticore-projects.github.io/sql/syntax.html

The railroad diagram is amazing!

kaikalur commented:
So just two changes needed to merge this PR: 1. fix the naming to 'Presto SQL Parser' and 2. move the formatter out to a separate package.

manticore-projects commented:
> So just two changes needed to merge this PR: 1. fix the naming to 'Presto SQL Parser' and 2. move the formatter out to a separate package.

Agreed. You will have it tomorrow. Thanks again for the guidance.

manticore-projects committed:
refactor: fix links and names

Signed-off-by: Andreas Reichel <andreas@manticore-projects.com>

manticore-projects commented:
Greetings.

I have removed all references to "manticore-projects" and/or "JSQLParser" and have applied "Presto SQL:2016 Parser" where possible.
I have also restored full Java 8 compatibility (which `airlift` silently dropped with version 100).
Gradle Build for the Parser module is fully function and I will stick with that one in the future.

Github Actions run CI and publish site.
Cheers!

manticore-projects committed:
doc: fix 2 more name references

Signed-off-by: Andreas Reichel <andreas@manticore-projects.com>

kaikalur commented on sphinx/src/contribution.rst:
```diff
@@ -70,7 +70,7 @@ Since **Presto SQL:2016 Parser** is built by JavaCC from a Token based Grammar,
     -- <OVERLAPS:"OVERLAPS"> is a Token, recently defined in the Grammar
     -- Although it is not restricted by the SQL Standard and could be used for Column, Table and Alias names
-    -- Explicitly white-listing OVERLAPS by adding it to the  RelObjectNameWithoutValue() Production will allow for parsing the following statement
+    -- Explicitly white-listing OVERLAPS by adding it to the  NonReservedKeywords() Production will allow for parsing the following statement
```
We try to use words like allow list and block list

manticore-projects replied:
What a time to be alive! I will change it.

manticore-projects replied:
What about ``white space characters`` though? :-))

kaikalur replied:
yeah so funny right - blank space?

manticore-projects committed:
doc: US compliant wording

Signed-off-by: Andreas Reichel <andreas@manticore-projects.com>

kaikalur reviewed (approved):
LGTM

kaikalur commented:
Thank you very much for doing this much needed setup cleanup!

kaikalur commented:
@mbasmanova do you think you can stamp this PR - it's simply adding a bunch of test/ci stuff with no changes to the grammar/ast side of things? I need to find some others who can do this now so that @manticore-projects and I can work on it and merge without delays.

manticore-projects committed:
doc: better RR Diagram integration

Signed-off-by: Andreas Reichel <andreas@manticore-projects.com>

manticore-projects committed:
doc: show Dependencies and Artefacts for both Maven and Gradle

manticore-projects committed:
refactor: Merge REPLACE into UPSERT

fixes #1706
